File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(number)
    raise RuntimeError unless number.positive?

    sum_of_factors = (1...number).select{ |factor| factor if (number % factor).zero? }.sum

    if sum_of_factors == number then 'perfect'
    elsif sum_of_factors > number then 'abundant'
    else 'deficient'
    end
  end
end

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError("Input must be a positive integer!")

    sum, i = 0, 1

    while i <= number / 2:
        if number % i == 0:
            sum += i

        i += 1

    if sum == number:
        return "perfect"
    if sum > number:
        return "abundant"
    if sum < number:
        return "deficient"
